src/ofonotracker.cpp

```diff
@@ -0,0 +1,335 @@
+#include <QTimer>
+#include <QDebug>
+
+/* ========================================================================= *
+ * String constants
+ * ========================================================================= */
+
+const auto DBUS_ERROR_NAME_HAS_NO_OWNER = QStringLiteral("org.freedesktop.DBus.Error.NameHasNoOwner");
+const auto OFONO_SERVICE = QStringLiteral("org.ofono");
+const auto OFONO_MANAGER_OBJECT = QStringLiteral("/");
+const auto OFONO_MANAGER_INTERFACE = QStringLiteral("org.ofono.Manager");
+const auto OFONO_MANAGER_GET_MODEMS = QStringLiteral("GetModems");
+const auto OFONO_MANAGER_MODEM_REMOVED = QStringLiteral("ModemRemoved");
+const auto OFONO_MANAGER_MODEM_ADDED = QStringLiteral("ModemAdded");
+const auto OFONO_MODEM_INTERFACE = QStringLiteral("org.ofono.Modem");
+const auto OFONO_MODEM_PROPERTY_CHANGED = QStringLiteral("PropertyChanged");
+const auto OFONO_MODEM_GET_PROPERTIES = QStringLiteral("GetProperties");
+const auto OFOFO_MODEM_PROPERTY_SERIAL = QStringLiteral("Serial");
+
+/* ========================================================================= *
+ * Custom D-Bus types
+ * ========================================================================= */
+
+struct ModemInfo
+{
+    ModemInfo(QString path, const QVariantMap &properties)
+        : m_objectPath(path)
+        , m_properties(properties)
+    {
+    }
+    ModemInfo(QString path)
+        : m_objectPath(path)
+    {
+    }
+    ModemInfo(const ModemInfo &that)
+        : m_objectPath(that.m_objectPath)
+        , m_properties(that.m_properties)
+    {
+    }
+    ModemInfo()
+    {
+    }
+    QDBusObjectPath m_objectPath;
+    QVariantMap m_properties;
+};
+Q_DECLARE_METATYPE(ModemInfo)
+
+QDBusArgument &operator<<(QDBusArgument &argument, const ModemInfo &modemInfo)
+{
+    argument.beginStructure();
+    argument << modemInfo.m_objectPath;
+    argument << modemInfo.m_properties;
+    argument.endStructure();
+    return argument;
+}
+const QDBusArgument &operator>>(const QDBusArgument &argument, ModemInfo &modemInfo)
+{
+    argument.beginStructure();
+    argument >> modemInfo.m_objectPath;
+    argument >> modemInfo.m_properties;
+    argument.endStructure();
+    return argument;
+}
+
+typedef QList<ModemInfo> ModemInfoList;
+Q_DECLARE_METATYPE(ModemInfoList)
+
+static void registerDBusTypes()
+{
+    static bool registered(false);
+    if (!registered) {
+        registered = true;
+        qDBusRegisterMetaType<ModemInfo>();
+        qDBusRegisterMetaType<ModemInfoList>();
+    }
+}
+
+/* ========================================================================= *
+ * OFonoTrackerPrivate
+ * ========================================================================= */
+
+class OFonoTrackerPrivate : public QObject, protected QDBusContext
+{
+    Q_OBJECT;
+public:
+    OFonoTrackerPrivate(bool synchronousInit, OFonoTracker *parent);
+    virtual ~OFonoTrackerPrivate();
+    QStringList imeiNumbers() const;
+private Q_SLOTS:
```

Comment:
Did we have some preference on `Q_SLOTS` vs. `slots` use? I didn't see anything in [coding conventions](https://docs.sailfishos.org/Develop/Apps/Coding_Conventions/#qt-coding-conventions). The latter seems to be almost exclusively used in this repository.

Btw, I would have a little bit more spacing here. Like an empty line before `public` and `private` and maybe also between functions and variables. Now it looks pretty crowded, but that might be just me.